docs/source/extension/extension_points.rst

```diff
@@ -951,10 +951,70 @@ Here is an example for making requests to the language server.
       const virtualDocument = adapter.virtualDocument;
 
       // Get the LSP connection of the virtual document.
-      const connection = manager.connections.get(document.uri);
+      const connection = manager.connections.get(virtualDocument.uri);
       ...
       // Send completion request to the language server
       const response = await connection.clientRequests['textDocument/completion'].request(params);
       ...
     }
   };
+
+Occasionally, LSP extensions include a CodeMirror extension to modify the code editor. In those cases, you can follow this example:
+
+.. code:: typescript
+
+  const renamePlugin: JupyterFrontEndPlugin<void> = {
+    id,
+    autoStart: true,
+    requires: [ILSPDocumentConnectionManager, ILSPFeatureManager, IWidgetLSPAdapterTracker],
+    activate: (app: JupyterFrontEnd, connectionManager: ILSPDocumentConnectionManager, featureManager: ILSPFeatureManager, tracker: IWidgetLSPAdapterTracker) => {
+      const FEATURE_ID = "rename_symbol";
+      const extensionFactory: EditorAdapter.ILSPEditorExtensionFactory = {
+        name: FEATURE_ID,
+        factory: (options) =>  {
+          const { editor, widgetAdapter } = options;
+
+          // Get the editor
+          const ceEditor: CodeEditor.IEditor | null = editor.getEditor();
+          if (!ceEditor) {
+            return null;
+          }
+
+          // Get the associated virtual document of the opened document
+          if (!widgetAdapter.virtualDocument) {
+            return null;
+          }
+
+          // Get the LSP connection of the virtual document.
+          const connection = connectionManager.connections.get(widgetAdapter.virtualDocument.uri);
+          if (!connection || !connection.provides('renameProvider')) {
+            return null;
+          }
+
+          // Create a CodeMirror extension that listens for double click, gets the
+          // selected code and makes a LSP request to rename it and prints the results.
+          const ext = EditorView.domEventHandlers({ dblclick: (e, view) => {
+            const range = ceEditor.getSelection();
+              const res = connection.clientRequests['textDocument/rename'].request({
+                newName: "test",
+                position: { line: range.start.line, character: range.start.column },
+                textDocument: { uri: widgetAdapter.virtualDocument!.uri }
+              });
+
+              res.then(value => {
+                console.debug(value);
+              }).catch(e => console.error);
+          }});
+
+          // Create a conditional CodeMirror extension.
+          return EditorExtensionRegistry.createConditionalExtension(ext);
```

Comment:
I guess `createConfigurableExtension` will be often useful, but for sake of having simple example I think it is better to use immutable vs conditional?

```suggestion
          // Wrap the CodeMirror extension in the extension registry object.
          return EditorExtensionRegistry.createImmutableExtension(ext);
```